client/v2/autocli/flag/file.go

```diff
@@ -0,0 +1,57 @@
+package flag
+
+import (
+	"context"
+	"encoding/hex"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"os"
+)
+
+type binaryType struct{}
+
+func (f binaryType) NewValue(_ context.Context, _ *Builder) Value {
+	return &fileBinaryValue{}
+}
+
+func (f binaryType) DefaultValue() string {
+	return ""
+}
+
+// fileBinaryValue is a Value that holds a binary file.
+type fileBinaryValue struct {
+	value []byte
+}
+
+func (f *fileBinaryValue) Get(protoreflect.Value) (protoreflect.Value, error) {
+	return protoreflect.ValueOfBytes(f.value), nil
+}
+
+func (f *fileBinaryValue) String() string {
+	return string(f.value)
+}
+
+func (f *fileBinaryValue) Set(s string) error {
+	var err error
+	var value []byte
+	// check if file exist
+	_, err = os.Stat(s)
+	if err == nil {
+		// open file at path s
+		value, err = os.ReadFile(s)
```

Comment:
Is it intended that we are just reading the file here but not attempting to decode its content?

Reply:
Yes, I think so. I think the idea is to allow users to pass data either by writing it to a file and passing `--flag=/path/to/file`, or by encoding it to hex and passing `--flag=<hex string>`. Whether or not this is a good idea is a separate question.

Reply:
Yes, @julienrbrt, that is what I understand from it. How do you think it should be?

@peterbourgon, if you don't mind, can you expand on whether you think it is or is not a good idea?

Reply:
I see. It feels a bit confusing that what you pass directly to a flag isn't the same as what you have in a file. I do understand the reason, though. If it's documented, I suppose it is fine.
Secondly, when I read the issue description: https://github.com/cosmos/cosmos-sdk/issues/13289, it seems like it should accept base64 *and* hex encoding, right now, the code behaves differently.

Reply:
There's no way to reliably determine if a flag's value represents a file name or a hex value.

```
echo 999999 > abc123
myprogram --flag=abc123
```

`abc123` is both a valid hex value _and_ a valid filename. Should `--flag` have a value of `abc123` or of `999999`? There is no way to know. Parsing is ambiguous.

How a flag's value is parsed should not be ambiguous. For this reason, a flag should not allow both filenames and/or (encoded forms of) raw data to be accepted, and I don't believe issues like #13288, or #13289, or etc. are sound. But I'm sure nobody really cares about these objections.

Reply:
Your argumentation makes sense, however #13288 reads a file when it has a specific extension, so it is better than here. We should probably enforce an arbitrary extension when reading a file here, too.

Reply:
> I see. It feels a bit confusing that what you pass directly to a flag isn't the same as what you have in a file. I do understand the reason, though. If it's documented, I suppose it is fine. Secondly, when I read the issue description: #13289, it seems like it should accept base64 _and_ hex encoding, right now, the code behaves differently.

Regarding the base 64, how will the user know how to read the value, I think it also applies to the file not being encoded as well at the end

Reply:
yes i could also just ensure it has a file extension

Reply:
You can't assume that all JSON files will have a `.json` extension.

Reply:
I changed it now it enforces that it uses an extension that is not a valid hex or base64

Reply:
I'm not sure that's a solution.

```go
if _, err = os.Stat(s); err == nil {
	if filepath.Ext(s) == "" {
 		return errors.New("file path must have an extension")
	}
```

If I create a valid file without an extension, like `abc`, wouldn't that break this code, and basically prevent the program from starting?

Reply:
Amm, it will just be an error when someone tries to pass a file without an extension to the CLI. The assumption would be to use a file; one will need a file extension.

Reply:
Okay, but then how is that assumption communicated to users?

Reply:
Good point, @peterbourgon ill make the error more verbose and add documentation.

Reply:
<details>
<summary>Code</summary>

```go
package main

import (
	"encoding/base64"
	"encoding/hex"
	"flag"
	"fmt"
	"log"
	"os"
)

type fileBinaryValue struct {
	value []byte
}

var _ flag.Value = (*fileBinaryValue)(nil)

func (f *fileBinaryValue) String() string {
	return string(f.value)
}

func (f *fileBinaryValue) Set(s string) error {
	if data, err := os.ReadFile(s); err == nil {
		f.value = data
		return nil
	}

	if data, err := hex.DecodeString(s); err == nil {
		f.value = data
		return nil
	}

	if data, err := base64.StdEncoding.DecodeString(s); err == nil {
		f.value = data
		return nil
	}

	return fmt.Errorf("invalid value %q", s)
}

func main() {
	log.SetFlags(0)
	var x fileBinaryValue
	fs := flag.NewFlagSet("myprogram", flag.ContinueOnError)
	fs.Var(&x, "x", "file name, or hex data, or base64 data")
	fs.Parse(os.Args[1:])
	log.Printf("x: %q", x.String())
}
```

```
$ go run main.go -x=610a
x: "a\n"

$ go run main.go -x="MTIzCg=="
x: "123\n"

$ echo abc def > file.txt
$ go run main.go -x=file.txt
x: "abc def\n"
```
</summary>

Reply:
It might be okay to prefer file, then hex, then base64 as this code suggests. For json, I actually think it's okay to not require `.json` for the filename because the only valid json that could also be a filename would be a number, true, false, or null. But whatever we choose it should be well documented and unsurprising. Tx submission should always prompt the user for confirmation by default if it's a concern that there could be files on disk that were intended to be just raw bytes or json.